### PR TITLE
Rem 538

### PR DESCRIPTION
**Pull request by Alinadimmokdad**
new page pos-pointOfSales

### CODE REVIEW

File: src/pages/pos-pointOfSales/forms/CashAccountForm.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+  const formik = useFormik({
+    enableReinitialize: true,
+    validateOnChange: true,
+    validationSchema: yup.object({
+      pOSUser: yup
+        .array()
+        .of(
+          yup.object().shape({
+            cashAccountId: yup.string().required('spId  is required')
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/CashAccountForm.js:28]: clear text

File: src/pages/pos-pointOfSales/forms/CashAccountForm.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        )
+        .required('userId array is required')
+    }),
+    initialValues: {
+      pOSUser: [
+        {
+          id: 1,
+          posId: recordId,
+          cashAccountId: '',
+
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/CashAccountForm.js:39]: remove space

File: src/pages/pos-pointOfSales/forms/CashAccountForm.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    validateOnChange: true,
+    validationSchema: yup.object({
+      pOSUser: yup
+        .array()
+        .of(
+          yup.object().shape({
+            cashAccountId: yup.string().required('spId  is required')
+          })
+        )
+        .required('userId array is required')
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/CashAccountForm.js:31]: remove text

File: src/pages/pos-pointOfSales/forms/CashAccountForm.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+                  isInactive: false
+                }
+              ]
+            })
+          }
+        })
+        .catch(error => {})
+  }
+  useEffect(() => {
+    getData()
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/CashAccountForm.js:143]: if  recordId before getData(

File: src/pages/pos-pointOfSales/forms/PointOfSalesForm.js [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+      plantId: yup.string().required(),
+      maxEndTime: yup.string().required(),
+      status: yup.string().required(),
+      plId: yup.string().required(),
+      siteId: yup
+        .string()
+        .nullable()
+        .test('siteId-required', 'Site ID is required when not an online store', function (value) {
+          const { onlineStore } = this.parent
+          console.log(onlineStore, 'onnn')
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/PointOfSalesForm.js:68]: remove console.log

File: src/pages/pos-pointOfSales/forms/PointOfSalesForm.js [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+          <ResourceComboBox
+            name='status'
+            label={labels.status}
+            datasetId={DataSets.POS_STATUS}
+            values={formik.values}
+            required
+            valueField='key'
+            displayField='value'
+            onChange={(event, newValue) => {
+              if (newValue) {
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/PointOfSalesForm.js:337]: can  u remove duplicate setFieldvalue

File: src/pages/pos-pointOfSales/forms/UsersForm.js [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+          email: '',
+          spName: ''
+        }
+      ]
+    },
+    onSubmit: values => {
+      postPOSUser(values)
+    }
+  })
+  console.log(formik.values.pOSUser)
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/UsersForm.js:50]: remove console.log

File: src/pages/pos-pointOfSales/forms/SalesPersonForm.js [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+    validateOnChange: true,
+    validationSchema: yup.object({
+      pOSUser: yup
+        .array()
+        .of(
+          yup.object().shape({
+            spId: yup.string().required('spId  is required')
+          })
+        )
+        .required('userId array is required')
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/SalesPersonForm.js:31]: remove text

File: src/pages/pos-pointOfSales/forms/SalesPersonForm.js [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+          isInactive: false,
+          spName: ''
+        }
+      ]
+    },
+    onSubmit: values => {
+      postPOSUser(values)
+    }
+  })
+  console.log(formik.values.pOSUser)
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/SalesPersonForm.js:48]: remove console.log

File: src/pages/pos-pointOfSales/forms/SalesPersonForm.js [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+      component: 'checkbox',
+      label: labels.isInactive,
+      name: 'isInactive'
+    }
+  ]
+
+  function getData() {
+    const defaultParams = `_posId=${recordId}`
+    var parameters = defaultParams
+    recordId &&
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/SalesPersonForm.js:99]: remove it and put if in use Effect

File: src/pages/pos-pointOfSales/forms/UsersForm.js [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+        ],
+        columnsInDropDown: [{ key: 'name', value: 'Name' }]
+      }
+    }
+  ]
+
+  function getData() {
+    const defaultParams = `_posId=${recordId}`
+    var parameters = defaultParams
+    recordId &&
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/forms/UsersForm.js:113]: remove recordId and put if in use Effect

File: src/pages/pos-pointOfSales/index.js [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+
+const PointOfSales = () => {
+  const { getRequest, postRequest } = useContext(RequestsContext)
+  const { platformLabels } = useContext(ControlContext)
+  const { stack } = useWindow()
+
+  async function fetchGridData(options = {}) {
+    const { _startAt = 0, _pageSize = 50 } = options
+
+    try {
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/index.js:25]: are u sure need try and catch here

File: src/pages/pos-pointOfSales/windows/PointOfSalesWindow.js [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+import SalesPersonForm from '../forms/SalesPersonForm'
+import CashAccountForm from '../forms/CashAccountForm'
+
+const PointOfSalesWindow = ({ recordId, labels, maxAccess }) => {
+  const [activeTab, setActiveTab] = useState(0)
+
+  const editMode = !!recordId
+
+  const [store, setStore] = useState({
+    recordId: recordId || null
```
Comment (omarZrayka23) [inline src/pages/pos-pointOfSales/windows/PointOfSalesWindow.js:15]: remove || null

**Comment by omarZrayka23** [comment]
Screenshot 2024-09-03 at 11 07 12 AM
save button is not working

File: src/pages/pos-pointOfSales/index.js (hunk 12 above)
Comment (Alinadimmokdad) [reply src/pages/pos-pointOfSales/index.js:25]: its better , if there is an error while fetching

**Comment by Ghinwa-Arab** [comment]
image

rename site to 'inventory site'

**Comment by Ghinwa-Arab** [comment]
image
image

This field should be in time format and has a drop down that shows current time as well as the time in 30-minute intervals.

File: src/pages/pos-pointOfSales/index.js [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+      const response = await getRequest({
+        extension: PointofSaleRepository.PointOfSales.qry,
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/index.js:27]: use page endpoint

File: src/pages/pos-pointOfSales/index.js [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+      return { ...response, _startAt: _startAt }
+    } catch (error) {}
+  }
+
+  const {
+    query: { data },
+    labels: _labels,
+    invalidate,
+    paginationParameters,
+    filterBy,
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/index.js:40]: filterBy not used, remove it

File: src/pages/pos-pointOfSales/index.js [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+    refetch,
+    access
+  } = useResourceQuery({
+    queryFn: fetchGridData,
+    endpointId: PointofSaleRepository.PointOfSales.qry,
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/index.js:45]: use page endpoint

File: src/pages/pos-pointOfSales/forms/PointOfSalesForm.js [hunk 17]
```diff
@@ -0,0 +1,375 @@
+import { Checkbox, FormControlLabel, Grid } from '@mui/material'
+import * as yup from 'yup'
+import toast from 'react-hot-toast'
+import FormShell from 'src/components/Shared/FormShell'
+import { ResourceIds } from 'src/resources/ResourceIds'
+import { useContext, useEffect, useState } from 'react'
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/PointOfSalesForm.js:6]: remove useState

File: src/pages/pos-pointOfSales/forms/PointOfSalesForm.js [hunk 18]
```diff
@@ -0,0 +1,375 @@
...
+import { DataSets } from 'src/resources/DataSets'
+import CustomNumberField from 'src/components/Inputs/CustomNumberField'
+
+const PointOfSalesForm = ({ labels, maxAccess, setStore, store }) => {
+  const { postRequest, getRequest } = useContext(RequestsContext)
+  const { platformLabels } = useContext(ControlContext)
+  const { recordId } = store
+
+  const invalidate = useInvalidate({
+    endpointId: PointofSaleRepository.PointOfSales.qry
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/PointOfSalesForm.js:30]: replace by page endpoint

File: src/pages/pos-pointOfSales/forms/PointOfSalesForm.js [hunk 19]
```diff
@@ -0,0 +1,375 @@
...
+        .string()
+        .nullable()
+        .test('siteId-required', 'Site ID is required when not an online store', function (value) {
+          const { onlineStore } = this.parent
+
+          return !!onlineStore || (!!value && value !== '')
+        })
+    }),
+    onSubmit: async obj => {
+      const response = await postRequest({
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/PointOfSalesForm.js:73]: add try catch

File: src/pages/pos-pointOfSales/forms/PointOfSalesForm.js [hunk 20]
```diff
@@ -0,0 +1,375 @@
...
+        formik.setFieldValue('recordId', response.recordId)
+        setStore(prevStore => ({
+          ...prevStore,
+          recordId: response.recordId
+        }))
+      } else toast.success(platformLabels.Edited)
+      invalidate()
+    }
+  })
+  const editMode = !!recordId
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/PointOfSalesForm.js:90]: replace !!recordId with !!formik.values.recordId

File: src/pages/pos-pointOfSales/forms/PointOfSalesForm.js [hunk 21]
```diff
@@ -0,0 +1,375 @@
...
+          })
+
+          formik.setValues({ ...res.record, onlineStore: Boolean(res.record.onlineStore) })
+        }
+      } catch (error) {}
+    })()
+  }, [])
+
+  return (
+    <FormShell form={formik} resourceId={ResourceIds.PointOfSales} maxAccess={maxAccess} editMode={editMode}>
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/PointOfSalesForm.js:108]: use vertLayout and grow

File: src/pages/pos-pointOfSales/forms/UsersForm.js [hunk 22]
```diff
@@ -0,0 +1,169 @@
...
+import { RequestsContext } from 'src/providers/RequestsContext'
+import * as yup from 'yup'
+import toast from 'react-hot-toast'
+import { VertLayout } from 'src/components/Shared/Layouts/VertLayout'
+import { Grow } from 'src/components/Shared/Layouts/Grow'
+import { ControlContext } from 'src/providers/ControlContext'
+import { PointofSaleRepository } from 'src/repositories/PointofSaleRepository'
+import { SaleRepository } from 'src/repositories/SaleRepository'
+
+const UsersForm = ({ store, labels, maxAccess, editMode }) => {
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/UsersForm.js:16]: editMode props not passed from window, remove it and declare it inside this form

File: src/pages/pos-pointOfSales/forms/UsersForm.js [hunk 23]
```diff
@@ -0,0 +1,169 @@
...
+          { from: 'recordId', to: 'spId' },
+          { from: 'name', to: 'spName' }
+        ],
+        columnsInDropDown: [{ key: 'name', value: 'Name' }]
+      }
+    }
+  ]
+
+  function getData() {
+    const defaultParams = `_posId=${recordId}`
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/UsersForm.js:110]: no need for this 
` const defaultParams = `_posId=${recordId}`
    var parameters = defaultParams`
remove and pass directly ``_posId=${recordId}`` to request parameters

File: src/pages/pos-pointOfSales/forms/UsersForm.js [hunk 24]
```diff
@@ -0,0 +1,169 @@
...
+    var parameters = defaultParams
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/UsersForm.js:111]: remove `recordId &&` before api request, and add if condition in useEffect block
`if(recordId) getData()`

File: src/pages/pos-pointOfSales/forms/UsersForm.js [hunk 25]
```diff
@@ -0,0 +1,169 @@
...
+    recordId &&
+      getRequest({
+        extension: PointofSaleRepository.PosUsers.qry,
+        parameters: parameters
+      })
+        .then(res => {
+          if (Array.isArray(res?.list) && res.list.length > 0) {
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/UsersForm.js:118]: no need for all this code
` if (Array.isArray(res?.list) && res.list.length > 0) {
            formik.setValues({
              pOSUser: res.list.map(({ ...rest }, index) => ({
                id: index,
                ...rest
              }))
            })
          } else {
            formik.setValues({
              posId: [
                {
                  id: 1,
                  posId: recordId,
                  userId: '',
                  spId: '',
                  email: '',
                  spName: ''
                }
              ]
            })
          }`
          
          u can only replace with, and it do the same job
            const modifiedList = res.list?.map((user, index) => ({
            ...user,
            id: index + 1
          }))

          formik.setValues({ pOSUser: modifiedList })"

File: src/pages/pos-pointOfSales/forms/SalesPersonForm.js [hunk 26]
```diff
@@ -0,0 +1,154 @@
...
+import { RequestsContext } from 'src/providers/RequestsContext'
+import * as yup from 'yup'
+import toast from 'react-hot-toast'
+import { VertLayout } from 'src/components/Shared/Layouts/VertLayout'
+import { Grow } from 'src/components/Shared/Layouts/Grow'
+import { ControlContext } from 'src/providers/ControlContext'
+import { PointofSaleRepository } from 'src/repositories/PointofSaleRepository'
+import { SaleRepository } from 'src/repositories/SaleRepository'
+
+const SalesPersonForm = ({ store, labels, maxAccess, editMode }) => {
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/SalesPersonForm.js:15]: editMode props not passed from window, remove it and declare it inside this form

File: src/pages/pos-pointOfSales/forms/CashAccountForm.js [hunk 27]
```diff
@@ -0,0 +1,172 @@
...
+import { RequestsContext } from 'src/providers/RequestsContext'
+import * as yup from 'yup'
+import toast from 'react-hot-toast'
+import { VertLayout } from 'src/components/Shared/Layouts/VertLayout'
+import { Grow } from 'src/components/Shared/Layouts/Grow'
+import { ControlContext } from 'src/providers/ControlContext'
+import { PointofSaleRepository } from 'src/repositories/PointofSaleRepository'
+import { CashBankRepository } from 'src/repositories/CashBankRepository'
+
+const CashAccountForm = ({ store, labels, maxAccess, editMode }) => {
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/CashAccountForm.js:15]: editMode props not passed from window, remove it and declare it inside this form

File: src/pages/pos-pointOfSales/forms/SalesPersonForm.js [hunk 28]
```diff
@@ -0,0 +1,154 @@
+import { useFormik } from 'formik'
+import { DataGrid } from 'src/components/Shared/DataGrid'
+import FormShell from 'src/components/Shared/FormShell'
+import { ResourceIds } from 'src/resources/ResourceIds'
+import { useContext, useEffect, useState } from 'react'
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/SalesPersonForm.js:5]: remove useState

File: src/pages/pos-pointOfSales/forms/SalesPersonForm.js [hunk 29]
```diff
@@ -0,0 +1,154 @@
...
+    const defaultParams = `_posId=${recordId}`
+    var parameters = defaultParams
+    recordId &&
+      getRequest({
+        extension: PointofSaleRepository.SalesPerson.qry,
+        parameters: parameters
+      })
+        .then(res => {
+          if (Array.isArray(res?.list) && res.list.length > 0) {
+            formik.setValues({
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/SalesPersonForm.js:105]: here u can do the same way I had mentioned for users form

File: src/pages/pos-pointOfSales/forms/SalesPersonForm.js [hunk 30]
```diff
@@ -0,0 +1,154 @@
...
+      component: 'checkbox',
+      label: labels.isInactive,
+      name: 'isInactive'
+    }
+  ]
+
+  function getData() {
+    const defaultParams = `_posId=${recordId}`
+    var parameters = defaultParams
+    recordId &&
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/SalesPersonForm.js:98]: remove ` recordId &&` and move it to useEffect

File: src/pages/pos-pointOfSales/forms/SalesPersonForm.js [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+    },
+    {
+      component: 'checkbox',
+      label: labels.isInactive,
+      name: 'isInactive'
+    }
+  ]
+
+  function getData() {
+    const defaultParams = `_posId=${recordId}`
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/SalesPersonForm.js:96]: remove     
const defaultParams = `_posId=${recordId}`
    var parameters = defaultParams
    
    and pass  `_posId=${recordId}` directrly to request parameters

File: src/pages/pos-pointOfSales/forms/CashAccountForm.js [hunk 32]
```diff
@@ -0,0 +1,172 @@
...
+          type: '',
+          isInactive: false
+        }
+      ]
+    },
+    onSubmit: values => {
+      postPOSUser(values)
+    }
+  })
+  console.log(formik.values.pOSUser)
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/CashAccountForm.js:49]: remove log

File: src/pages/pos-pointOfSales/forms/CashAccountForm.js [hunk 33]
```diff
@@ -0,0 +1,172 @@
...
+      }
+    },
+    {
+      component: 'checkbox',
+      label: labels.isInactive,
+      name: 'isInactive'
+    }
+  ]
+
+  function getData() {
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/CashAccountForm.js:109]: refactor getData as mentioned for previous forms

File: src/pages/pos-pointOfSales/forms/CashAccountForm.js [hunk 34]
```diff
@@ -0,0 +1,172 @@
...
+        })
+        .catch(error => {})
+  }
+  useEffect(() => {
+    if (recordId) {
+      getData()
+    }
+  }, [recordId])
+
+  console.log('DataGrid value:', formik.values.pOSUser)
```
Comment (Ghinwa-Arab) [inline src/pages/pos-pointOfSales/forms/CashAccountForm.js:147]: remove log

File: src/repositories/PointofSaleRepository.js [hunk 35]
```diff
@@ -6,9 +6,24 @@ export const PointofSaleRepository = {
...
     set: service + 'setUSR',
     del: service + 'delUSR',
-    page: service + 'pageUSR'
+    page: service + 'pageUSR',
+    set2: service + 'set2USR'
   },
   PosUsersPOS: {
     qry: service + 'qryPOS'
+  },
+  PointOfSales: {
```
Comment (Ghinwa-Arab) [inline src/repositories/PointofSaleRepository.js:15]: merge all endpoints you add to the above block directly, already POS endpoint exists

**Comment by NoraBitarSM** [comment]
in edit mode:
currency is enabled
inventory site is disabled

in users tab, sales person combo show Ref and Name in the dropdown and on selection

in sales person tab, sales person combo show Ref and Name in the dropdown and on selection

in cash account tab, cash account combo, show Ref and Name in the dropdown and on selection

**Pull request merged**